bip-taro.mediawiki

```diff
@@ -470,6 +594,67 @@ other normal script-path scripts within the tapscript tree.
 The above example creates only a single asset. It's possible to also create
 several assets within a single output, thereby batching asset creation.
 
+===Asset Burning===
+
+Given the requirement that Taro commitments be unique within the context of a
+tapscript commitment, Taro assets can also be ''provably burned''.
+
+In order to burn an asset, the holder of the asset can simply ''remove'' the
+Taro asset leaf from their asset tree.
+
+In order to prove to a 3rd party that an asset has been burnt, the holder of
+said asset must reveal the following information:
+* A non-inclusion proof rooted at their asset tree that proves that asset in question is no longer committed to.
+* A proof for each other top-level output in the burning transaction that asserts the following:
+** For each output, that doesn't commit to the asset tree the non-inclusion proof is rooted at, the first depth of the control block leaf reveal must be presented:
+*** If the output key was created via BIP 86, then simply verify the mapping based on the internal key. In this case there is no script tree being committed to.
+*** If the proof is of depth 0, meaning only a single value is committed to, then assert the item isn't a valid Taro asset root
+*** If the proof is of depth 1, then assert that either both elements are tap branches, or one of the items is a leaf that isn't a valid Taro asset root.
+
+Asset burning and verification can be expressed via the following algorithms:
+
+<source lang="python">
+burn_asset(asset_id: AssetID, asset_script_key: SchnorrKey, taro_root: MerkleSumSMT) -> (PkScript, SMTProof):
+    # Obtain the asset SMT sub-tree within the main taro commitment.
+    asset_smt = taro_root.fetch_asset_tree(asset_key)
+
+    # Next, delete the asset leaf and propagate that change to generate a new root.
+    new_asset_root = asset_smt.delete_leaf(asset_id, asset_script_key).merkle_up()
+
+    # We'll now replace the asset ID sub tree in the main taro tree.
+    new_taro_root = asset_smt.replace_key(asset_id).merkle_up()
+
+    # Now that we have the new root, we'll use a fresh internal key and generate the output for it.
+    internal_key = new_internal_key()
+    taro_output = taproot_output_script(key=internal_key, leaves=[new_taro_root])
+
+    # The final step is to generate a non-inclusion proof to show the asset is no being committed to.
+    asset_burn_proof = gen_non_inclusion_proof(new_taro_root, new_asset_root, asset_id, asset_script_key)
```

Comment:
To generate a non inclusion proof, shouldn't `gen_non_inclusion_proof` require the asset that's being burned to be an argument?
Victor: Prove to me you don't commit to those 20 beefbux
Peggy: ah here is a proof, that's uses those 20 beefbux as a predicate, that if i could claim them, they would be here, and as I show you, they are not.

Reply:
The asset ID and script key are passed in.